### ISSUE TITLE
optional input_fields sent as nil

### ISSUE THREAD

**Issue by note89** [jahewson/elm-graphql#26]
When haveing optional fields like this

```
    def input_fields do
       %{
         notification_id: string_non_null,
         user_id: string,
       }
    end
```

you have to send along `Nothing` for `user_id` if you are not interested. 
I think that the default behaviour should be that inputs with `Nothing` should not be sent to the backend.
In relay i just skip the field and i have checks like this on the backend. 

```
        user_id =
            if !Map.has_key?(input, "user_id") do
                info[:root_value][:current_user].id
            else
                [_type, id] = Node.from_global_id(input["user_id"])
                id
            end
```

But this fails if `"user_id" => nil ` is sent to the backend.

**Comment by jahewson** [comment]
I think your backend code might be the problem. The [GraphQL spec](https://facebook.github.io/graphql/#sec-Input-Objects) has this to say (emphasis mine):

> ### Input Coercion
The input to an input object should be an unordered map, otherwise an error should be thrown. The result of the coercion is an unordered map, with an entry for each input field, whose key is the name of the input field. The value of an entry in the coerced map is the result of input coercing the value of the entry in the input with the same key; **if the input does not have a corresponding entry, the value is the result of coercing null**. The input coercion above should be performed according to the input coercion rules of the type declared by the input field.

So you should be treating missing keys as equivalent to the key existing but having the value null. There is no semantic difference between a value of null and a missing key.

**Comment by note89** [comment]
But your sending unesseary data across the wire.  When using Relay the keys are not sent along (same backend)

**Comment by jahewson** [comment]
That's true, the `null` values are not necessary, I could re-open this issue as a feature request if you like?

But just to be clear, there's no bug as your backend should treat missing keys and null identically.

**Comment by note89** [comment]
Tnx :) I will fix my backend.

**Comment by jahewson** [comment]
Reopened as a feature request.

**Comment by note89** [comment]
I have a question about this, when creating update functionality how would one solve this case. 
You have 10 fields and non of them can be null. 
Its okey to just update one of them. 
```
key1: value
```
here we should only update `key1` and don't throw a error.
But if `key1` was send as 
```
key: null
```
then we should complain. 

but now if we send all keys everytime 

```
key1: value,
key2: null,
key3: null,
...
```
We cannot complain on `null` but how do we solve this then ?

**Comment by note89** [comment]
Actually, the spec was changed on `Oct 16` last year. So the backend should differentiate between missing and explicit `null` from now on.
